src/controller/CHIPDeviceController.cpp

```diff
@@ -1178,6 +1184,298 @@ void DeviceCommissioner::OnSessionEstablished()
+{
+    VerifyOrReturnError(mState == State::Initialized, CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrReturnError(mDeviceBeingPaired < kNumMaxActiveDevices, CHIP_ERROR_INCORRECT_STATE);
+
+    Device * device = &mActiveDevices[mDeviceBeingPaired];
+
+    ByteSpan certDeclaration;
+    ByteSpan attestationNonce;
+    ByteSpan timestamp;
+    ByteSpan firmwareInfo;
+    ByteSpan vendorReserved1;
+    ByteSpan vendorReserved2;
+    ByteSpan vendorReserved3;
+    ByteSpan vendorReserved4;
+
+    ByteSpan * element_array[] = { &certDeclaration, &attestationNonce, &timestamp,       &firmwareInfo,
+                                   &vendorReserved1, &vendorReserved2,  &vendorReserved3, &vendorReserved4 };
+
+    System::PacketBufferTLVWriter tlvWriter;
+    TLV::TLVType outerContainerType = TLV::kTLVType_NotSpecified;
+    System::PacketBufferHandle attestationElementsTbs;
+
+    TLV::TLVReader tlvReader;
+    TLV::TLVType containerType = TLV::kTLVType_Structure;
+    int i                      = 0;
+
+    tlvReader.Init(attestationElements.data(), static_cast<uint32_t>(attestationElements.size()));
+    ReturnErrorOnFailure(tlvReader.Next(containerType, TLV::AnonymousTag));
+    ReturnErrorOnFailure(tlvReader.EnterContainer(containerType));
+    do
+    {
+        if (!TLV::IsContextTag(tlvReader.GetTag()))
+        {
+            continue;
+        }
+        const uint8_t * data = nullptr;
+        ReturnErrorOnFailure(tlvReader.GetDataPtr(data));
+        *element_array[i++] = ByteSpan(data, tlvReader.GetLength());
+    } while (tlvReader.Next() == CHIP_NO_ERROR);
+
+    // Step f
+    // Retrieve attestation challenge
+    PeerConnectionState * state = mSessionMgr->GetPeerConnectionState(
+        { mPairingSession.PeerConnection().GetPeerNodeId(), mPairingSession.GetPeerKeyId(), mAdminId });
+    VerifyOrReturnError(state != nullptr, CHIP_ERROR_NOT_CONNECTED);
+    ByteSpan attestationChallenge = state->GetSecureSession().GetAttestationChallenge();
+
+    // Create attestation-elements-tbs
+    attestationElementsTbs = System::PacketBufferHandle::New(System::PacketBuffer::kMaxSize);
+    VerifyOrReturnError(!attestationElementsTbs.IsNull(), CHIP_ERROR_NO_MEMORY);
+
+    tlvWriter.Init(std::move(attestationElementsTbs));
+    outerContainerType = TLV::kTLVType_NotSpecified;
+    ReturnErrorOnFailure(tlvWriter.StartContainer(TLV::AnonymousTag, TLV::kTLVType_Structure, outerContainerType));
+    ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(0), attestationElements));
+    ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(1), attestationChallenge));
+    ReturnErrorOnFailure(tlvWriter.EndContainer(outerContainerType));
+    ReturnErrorOnFailure(tlvWriter.Finalize(&attestationElementsTbs));
+
+    // Step e
+    P256ECDSASignature deviceSignature;
+    ReturnErrorOnFailure(deviceSignature.SetLength(signature.size()));
+    memcpy(deviceSignature, signature.data(), signature.size());
```

Comment:
Need to bound-check that incoming `signature` is of expected size (kP256_ECDSA_Signature_Length_Raw), otherwise we will get stack overrun from `deviceSignature` down.

Reply:
The SetLength method above already checks if the new length is < capactity
+ the ECDSA_validate_hash_signature method already does the suggested check.
Therefore,  I think we shouldn't need to add that extra bound-check

Reply:
This is done after the fact. You are right that:

`ReturnErrorOnFailure(deviceSignature.SetLength(signature.size()));` 

does avoid what I said re: overrunning from the memcpy. The fact that it was not immediately obvious (without looking at impl), means that it's a risky pattern. Suggest adding a comment line above that SetLength will fail if signature doesn't fit.

Furthermore, the schema validation of incoming data in cluster command handler should validate the exact size of signature. This is not done. That is where the error should have occured. Otherwise we are deep in the guts of this procedure and it aborts when the failure should have been much before any of this code is run, since the signature would have been larger than allowed in the initial payload.

Reply:
added comment to state that the AttestationInfo will fail to be validated if the signature doesn't fit